### DISCUSSION TITLE
Optimizing element-wise operations

### DISCUSSION THREAD

**Discussion by cuihantao** [rust-ndarray/ndarray#1240] [Q&A] [answered]
Hi,

I'm new to Rust and ndarray and have been tinkering with element-wise array operations. Below is the code snippet to multiply five arrays and write to `arr_out`.

The code in the `for` loop takes 783 ns, and the commented-out line takes 1.24 µs. I checked the generated asm for the `for` loop; it does not use `vmulpd` but instead uses `mulsd`. These vectors are of the same length, but I'm not sure how to inform the compiler of it; adding `assert!` worsens the performance. 

I also tried `itertools.izip!` but it's not as fast as the for loop. What is the best way to do element-wise operations? My expressions will have many variables so a brief syntax is preferred, but performance is the top priority. 

```rust
pub fn mul_1d_inplace<'a>(arr_out: &'a mut Array1::<f64>,
                          arr1: &'a Array1::<f64>,
                          arr2: &'a Array1::<f64>,
                          arr3: &'a Array1::<f64>,
                          arr4: &'a Array1::<f64>,
                          arr5: &'a Array1::<f64>,
                          ) -> Option<f64> {
    for i in 0..arr1.len() {
        arr_out[i] = arr1[i] * arr2[i] * arr3[i] * arr4[i] * arr5[i];
    }

    // *arr_out = arr1 * arr2 * arr3 * arr4 * arr5;  // slower

    None
}
```

**Comment by adamreichold** [comment, marked as answer]
Please have a look at the [`Zip`](https://docs.rs/ndarray/latest/ndarray/struct.Zip.html) type and for convenience the [`azip!`](https://docs.rs/ndarray/latest/ndarray/macro.azip.html) macro.

**Reply by cuihantao** [reply]
That really helps. With `azip!`, the assembly now uses `mulpd`.
